### PR TITLE
travis & docker overhaul

### LINKED ISSUE #268: Travis updates
- use stages (https://docs.travis-ci.com/user/build-stages/)
    + use conditionals for deployment, maybe allow Mac builds to fail https://docs.travis-ci.com/user/conditional-builds-stages-jobs/
- may help with #167
- change default `dist` to 16.04 Xenial (https://docs.travis-ci.com/user/trusty-to-xenial-migration-guide/) or even 18.04 Bionic (https://docs.travis-ci.com/user/reference/bionic/) instead of Trusty 14.04 (but perhaps retain one trusty build)

### LINKED ISSUE #281: fix docker ccache
Current builds (https://travis-ci.org/CCPPETMR/SIRF-SuperBuild) take about 4 times longer on `docker` because `.ccache` is no longer being used properly. Needs to be updated/re-fixed.

Also we should be re-building without caching `ccppetmr/sirf:core` (and also maybe not caching `ccppetmr/sirf:base`)

### PR DESCRIPTION
**Pull request by casperdcl**
- [x] fixes #281 docker caching
  + [x] revisit docker image caching
  + [x] refix docker `ccache`
- [x] fixes #268 travis upgrade
  + [x] use build stages
  + [x] use conditionals for deployment
  + [x] allow Mac builds to fail
  + [x] change default to `dist: bionic` (18.04)
  + [x] add `dist: xenial` (16.04)
  + [x] keep one `dist: trusty` (14.04)
  + [x] restructure jobs matrix to avoid duplication and unnecessary commands
- may help with #167

### CODE REVIEW

**Comment by casperdcl** [comment]
[diff without whitespace](https://github.com/CCPPETMR/SIRF-SuperBuild/pull/284/files?diff=unified&w=1)

File: docker/user_sirf-ubuntu.sh [hunk 1]
```diff
@@ -3,11 +3,15 @@
 set -ev
 INSTALL_DIR="${1:-/opt}"
 # SIRF
-git clone https://github.com/CCPPETMR/SIRF-SuperBuild --recursive -b master $INSTALL_DIR/SIRF-SuperBuild
+git clone https://github.com/CCPPETMR/SIRF-SuperBuild --recursive -b docker-ccache $INSTALL_DIR/SIRF-SuperBuild
```
Comment (KrisThielemans) [inline docker/user_sirf-ubuntu.sh]: seems incorrect

File: .travis.yml [hunk 2]
```diff
@@ -1,143 +1,321 @@
...
-sudo: required  # for docker
-services:
-- docker
+dist: bionic  # Ubuntu 18.04 LTS
 cache:  # shared between builds
 - ccache
 - pip
+stages:
+# - test  # TODO: uncomment before merge
+- name: test  # TODO: remove before merge
```
Comment (KrisThielemans) [inline .travis.yml]: I guess the `TODO` should have been removed as well (also below)

File: docker/user_sirf-ubuntu.sh (hunk 1 above)
Comment (casperdcl) [reply docker/user_sirf-ubuntu.sh]: yes I pushed a commit to master immediately after merge reverting this. Just want to force refresh master trvis caches. Will take a while to build but everything (except of course osx) should pass.

File: .travis.yml (hunk 2 above)
Comment (casperdcl) [reply .travis.yml]: yes, was part of merge commit

**Comment by casperdcl** [comment]
hmm before there's: https://github.com/CCPPETMR/SIRF-SuperBuild/blob/ea19bf415a2e7b00696752ee751859d65ca85f17/SuperBuild.cmake#L61 and after: https://github.com/CCPPETMR/SIRF-SuperBuild/blob/ea19bf415a2e7b00696752ee751859d65ca85f17/SuperBuild.cmake#L262 is that correct?
Will try reverting these changes later in another branch once all the caches have finished rebuilding and see if it makes a difference.

**Comment by KrisThielemans** [comment]
https://github.com/CCPPETMR/SIRF-SuperBuild/blob/ea19bf415a2e7b00696752ee751859d65ca85f17/SuperBuild.cmake#L61
not sure what that does, but I believe @rijobro had to put it in for the library names. can't see what would be wrong with it.

https://github.com/CCPPETMR/SIRF-SuperBuild/blob/ea19bf415a2e7b00696752ee751859d65ca85f17/SuperBuild.cmake#L262
> is that correct?

why wouldn't it be?  somewhat weirdly, all projects currently depend on `PRIMARY_PROJECT_NAME`, but that doesn't have anything to do with `mark_as_superbuild`

**Pull request merged**